File: 438.find-all-anagrams-in-a-string.java
/*
 * @lc app=leetcode id=438 lang=java
 *
 * [438] Find All Anagrams in a String
 */

// @lc code=start

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        if (s.length() < p.length()) {
            return new ArrayList<Integer>();
        }
        Map<Character, Integer> expected = new HashMap<>();
        for (char ch : p.toCharArray()) {
            expected.put(ch, expected.getOrDefault(ch, 0) + 1);
        }
        Map<Character, Integer> win = new HashMap<>();
        int ix = 0;
        char[] chars = s.toCharArray();
        while (ix != p.length()) {
            win.put(chars[ix], win.getOrDefault(chars[ix], 0) + 1);
            ix += 1;
        }
        List<Integer> ret = new ArrayList<>();
        if (win.equals(expected)) {
            ret.add(ix - p.length());
        }
        while (ix != chars.length) {
            win.put(chars[ix], win.getOrDefault(chars[ix], 0) + 1);
            win.put(chars[ix - p.length()], win.get(chars[ix - p.length()]) - 1);
            win.remove(chars[ix - p.length()], 0);
            if (win.equals(expected)) {
                ret.add(ix - p.length() + 1);
            }
            ix += 1;
        }
        return ret;
    }
}
// @lc code=end

File: 438.find-all-anagrams-in-a-string.cpp
/*
 * @lc app=leetcode id=438 lang=cpp
 *
 * [438] Find All Anagrams in a String
 */

// @lc code=start

#include <vector>
#include <string>
#include <array>

using namespace std;

class Solution {
public:
    using table_type = array<int, 26>;
    vector<int> findAnagrams(string s, string p) {
        if (s.size() < p.size()) {
            return {};
        }
        table_type pt = buildTable(p);
        table_type st;
        auto l = 0, r = 0;
        while (r - l != p.size()) {
            st[s.at(r) - 'a'] += 1;
            r += 1;
        }
        vector<int> ret;
        while (r <= s.size()) {
            if (isAnagram(st, pt)) {
                ret.push_back(l);
            }
            st[s.at(l) - 'a'] -= 1;
            l += 1;
            if (r == s.size()) {
                break;
            }
            st[s.at(r) - 'a'] += 1;
            r += 1;
        }
        return ret;
    }
    table_type buildTable(const string &s) {
        table_type t;
        for (auto ch: s) {
            t[ch - 'a'] += 1;
        }
        return t;
    }
    bool isAnagram(const table_type &a, const table_type &b) {
        table_type t;
        copy(a.begin(), a.end(), t.begin());
        transform(a.begin(), a.end(), b.begin(), t.begin(), [](const auto x, const auto y) {
            return x - y;
        });
        return all_of(t.begin(), t.end(), [](const auto x) {
            return x == 0;
        });
    }

};
// @lc code=end
